### DISCUSSION TITLE
I/O pins

### DISCUSSION THREAD

**Discussion by OldSurferDude** [Dan-in-CA/SIP#295] [Q&A] [answered]
I am going to have my own I/O pin control because I'm using a cross-bar to turn on valves, That means that two relays have to turn on to turn on a station's valve.

Which I/O pins does SIP use?  In my thrashing around, I thought that I had come across a setting that disabled the use of pins by SIP.

For what does SIP use the I2C bus?  expansion boards?  I see that it can be disabled.

Now the question comes, if SIP isn't using any I/O pins, could it run on any linux box?

Thanks for your consideration.

OSD

**Comment by Dan-in-CA** [comment]
When you say "two relays have to turn on to turn on a station's valve' does that mean 2 relays per valve or can you use a master relay for all stations?

SIP uses different IO pins depending on the version of Pi it is running on. There are also 2 different pin numbering systems. GPIO (BMC) numbering based on the processor chip pin-out and board numbering based on the header on the PI. A good cross reference is:
https://pinout.xyz/#

On the latest Pi models with 40 pin headers SIP uses GPIO (BMC) pins 27,  4, 17, 22, 14, and 15.

A number of plugins use other pins.

SIP variable: gv.use_gpio_pins controls SIP's use of GPIO pins, default = True, Set = False to disable. This would need to be set in software. However there is a new Node-RED plugin under development that can allow setting most SIP variables. If you are not familiar with node-red:
https://nodered.org/

SIP only uses the I2C bus for add-ons such as a real time clock or expansion boards like those used in the pcf857x_plugin. You might find that one interesting:
https://gitlab.com/seventer/pcf857x_plugin_sip/-/blob/main/doc/pcf857x-doc.md

Yes SIP can be run on any Linux box but of course the Pi uses the least amount of power considering it is intended to be running 24/7/365 for years.

Dan

**Reply by OldSurferDude** [reply]
Thanks Dan

So I would modify the file gv.py?  I assume that since python is an interpreted language that this program is run every time SIP is run.

You only note 6 pins.  The default in SIP is 8 stations. The other 2? Where do I find what plugins use what pins?

Think of a  "crossbar" as a set of stiff bare wires, some, all parallel, in a plane and some in another plane, also all parallel, but perpendicular to the wires the first plane.  A wire from each plane is moved towards the other plane until these wires make contact.  If the crossbar is 5 x 5 it is a 25 way switch.

In my case, I have 5 relays with the NO terminal connected to GND and the other 5 relays with NO connected to +12V.  There are 10 wires that go to the stations.  I then connect the solenoid of the station to one of the GND and one of the 12V wires.  When I turn  on the two corresponding relays the solenoid is energized opening the valve.

My solenoids are DC and have two diodes, flyback and reverse current prevention.  The crossbar concept won't work with AC)  My irrigation is gravity fed and I need to use "zero pressure" valves.  Usual irrigation valves require a constant 20 PSI to operate.  The max pressure in my system is at the bottom of the hill and is about 12 PSI.

I designed this system several years ago and a 25 zone controller was about $600.  I have recently found some ESP8266 relays for about $4. I considering a complete redesign using one for each zone.  Which feeds well into SIP :)  They require a power supply, but I already have 10 wires at each station, and with the new design, I only need 2.

You make a good point about the power usage, an RPi is ~10W.  My linux box is ~50W.  I have two considerations.  I am going to be running about six process, SIP being one of them.  The RPi can do amazing things, but I'm uncomfortable pushing it.  Also, these days, a bare RPi is about $150.  I can get a [refurbished computer Core i5-4570 3.2GHz 8GB 500GB DVDRW)](https://www.amazon.com/gp/product/B075WYN1JL/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&th=1) for about $100.

Food for thought

Thanks for your help!

OSD

**Comment by Dan-in-CA** [comment, marked as answer]
The 6 pins SIP uses by default are: 
4 to control a shift register system (The original design SIP was developed from)
1 for a rain sensor input
1 for a single relay output to control external hardware. 

The relay_16 plugin can allow up to 16 stations to be connected directly to the IO pins on a Pi. See the image below for details.
![Relay_16-connect](https://user-images.githubusercontent.com/4706639/207438031-94b871fd-2108-409c-82c8-ac1ac975aa6a.png)

The pcf857x_plugin only uses 4 pins (I2C) but allows connecting up to 64 stations with ~$2 expansion boards (8 per board).

**Reply by OldSurferDude** [reply]
Thanks Dan,

I so focused on my implementation that I blinded myself to other possibilities.

OSD
